### DISCUSSION TITLE
Should the Plugin or Attacher type be deprecated?

### DISCUSSION THREAD

**Discussion by remcohaszing** [unifiedjs/unified#136] [General] [answered]
I noticed the [`Plugin`](https://github.com/unifiedjs/unified/blob/main/types/ts4.0/index.d.ts#L225) type is an exact alias of [`Attacher`](https://github.com/unifiedjs/unified/blob/0f3b92833fd910c21edef423ad5891af7f11a82e/types/ts4.0/index.d.ts#L294-L297). I wonder if this is done purposely or if one should be removed (or deprecated for backwards compatibility).

**Comment by ChristianMurphy** [comment, marked as answer]
> I wonder if this is done purposely

That goes all the way back to https://github.com/unifiedjs/unified/pull/43, @Rokt33r may be able to comment more on the intent.
My educated guess, is that it reflects that the documentation for Unified uses both terms somewhat interchangeably, and that the aliases are used to match the documentation.

---

> if one should be removed (or deprecated for backwards compatibility).

:thinking: Out in the community, from what I've seen, `Plugin` is what I've used most and seen used most.
I'd be open to deprecating `Attacher` since it isn't frequently used.
Though I can also see another perspective, where it aligns with the documentation, and doesn't seem to be doing any harm (that I have seen), it might be okay to leave it be as well. :man_shrugging: 

Thoughts? :thought_balloon:

**Reply by wooorm** [reply]
I can imagine a future where another type is also added to `Plugin`, some function that isn’t an attacher, or an object with other fields that describes a plugin. That would not be an `Attacher`, but would add to `Plugin`?

**Reply by remcohaszing** [reply]
I usually prefer having one clear way of how to use something. I believe having multiple options to do the exact same thing causes ambiguity and code inconsistency (this applies to programming in general).

That having said, I don’t consider this to be a big issue. I was just unsure which one to use.

**Reply by ChristianMurphy** [reply]
Type changes for the next major (https://github.com/unifiedjs/unified/commit/0e8f61134e948962be86523cd0671b792217dcd2 and https://github.com/unifiedjs/unified/pull/141) are falling into place.
If you're interested in opening a PR changing the typings or removing one of the types, now could be a good time, as it would likely be a semver major change.

**Reply by remcohaszing** [reply]
I have been a bit busy lately, but I do want to fix this soon. Should I just remove `Attacher` then? IMO `Plugin` seems more appropriate.

**Reply by wooorm** [reply]
I’m open to it. But I can then still see https://github.com/unifiedjs/unified/discussions/136#discussioncomment-703928 happening in the future, where it’s split out again (which we could discuss then)

**Reply by remcohaszing** [reply]
This is really minor. If @wooorm prefers to keep both the `Plugin` and `Attacher` type we can just keep them and close this discussion.

**Reply by wooorm** [reply]
Given that we’re all a bit 🤷‍♂️ on whether or not to do this and unified is already seeing quite a few (good) breaking changes at the moment that are causing confusion and updating troubles, even though it’s probably a good idea, I think it’s best to *not* change it to not rock the boat further.
Okay to table it and keep it as is?

**Reply by ChristianMurphy** [reply]
`Attacher` has been deprecated in 10 https://github.com/unifiedjs/unified/blob/4919517cf6bd25e801322c3e0ba817753435e4fd/index.d.ts#L394-L399
